src/cript/nodes/core.py

```diff
@@ -72,3 +73,73 @@ def json(self):
         from cript.nodes.util import NodeEncoder
 
         return json.dumps(self, cls=NodeEncoder)
+
+    def find_children(self, search_attr:dict, recursion_depth:int=-1) -> List:
+        """
+        Finds all the children in a given tree of nodes (specified by its root),
+        that match the criteria of search_attr.
+        If a node is present multiple times in the graph, it is only once in the search results.
+
+        recursion_dept: Max depth of recursion into the tree. Helpful if circles are expected. -1 specifies no limit
+
+        search_attr: dict
+           Dictionary that specifies which JSON attributes have to be present in a given node.
+           If an attribute is a list, it it is suffiecient if the specified attributes are in the list,
+           if others are present too, that does not exclude the child.
+
+           Example: search_attr = `{"node": "Parameter"}` finds all "Parameter" nodes.
+                    search_attr = `{"node": "Algorithm", "parameter": {"name" : "update_frequency"}}` finds all "Algorithm" nodes, that have a parameter "update_frequency".
+                                  Since parameter is a list an alternative notation is ``{"node": "Algorithm", "parameter": [{"name" : "update_frequency"}]}` and Algorithms are not excluded they have more paramters.
+                    search_attr = `{"node": "Algorithm", "parameter": [{"name" : "update_frequency"}, {"name" : "cutoff_distance"}]}` finds all algoritms that have a parameter "update_frequency" and "cutoff_distance".
+
+        """
+        def is_attr_present(node:BaseNode, key, value):
+            """
+            Helper function that checks if an attribute is present in a node.
+            """
+
+            attr_key = asdict(node._json_attrs).get(key)
+
+            # To save code paths, I convert non-lists into lists with one element.
+            if not isinstance(attr_key, list):
+                attr_key = [attr_key]
+            if not isinstance(value, list):
+                value = [value]
+
+            number_values_found = 0
+            for v in value:
+                if v in attr_key:
+                    number_values_found += 1
+
+                for attr in attr_key:
+                    if isinstance(attr, BaseNode) and isinstance(v, dict):
+
+                        if len(attr.find_children(v, 0)) > 0:
+                            number_values_found += 1
+                            break
+            return number_values_found == len(value)
+
+
+        found_children = []
+
+        # Handle self
+        found_attr = 0
+        for key, value in search_attr.items():
+            if is_attr_present(self, key, value):
+                found_attr += 1
+        if found_attr == len(search_attr):
+            found_children += [self]
+
+        # Handle recursion
+        if recursion_depth != 0:
+            for field in self._json_attrs.__dataclass_fields__:
```

Comment:
I think this works perfectly fine for now, and it is exactly what we need to keep the code simple and good for the first version. In the future we probably would want to implement some other kind of search algorithms to make this search faster and more efficient. Like if we could make it into binary search then we could reduce the time complexity all the way down to `O(Logn)` which is just beautiful! But for first version I do not want to make this complicated at all.

Reply:
Yeah, it is possible.
We would have to make the nodes sortable (so implementing a > condition). It is certainly possible by using a hash value from the json and make those sortable.
However, the search condition must understand this order then too. I don't think that would work with hashes though. No sure. Maybe someone smarter then knows something here.
The trouble is that, the specified conditions can be partial, which also rules out hash table search.

Maybe it is possible to implement some variation of bucket sort, where we sort the nodes in based on present attributes.
That would be constant in the number of nodes, better then log(N) and linear in the number of search criterions.
That is probably the best way to do it. But maybe something for the back end, rather then SDK.
It would also require a separate search structure that is build as nodes are loaded into the SDK, which is way too much overhead for the SDK.

Reply:
yeah I agree, lets come back to it in future versions